Implement lightning.kachow!

JayOrten opened the pull request:
This PR implements PyTorch Lightning functionality

The major contributions of this PR are:
- train_model_lightning.py: implements a training script utilizing PyTorch Lightning
- dataset.py: contains the PyTorch Lightning DataModule class
- models.py: contains the Torch Lightning module model classes and the HuggingFace module model classes
- generation.py: generation script for generating text with a trained model
- inference.py: inference script for running trained model on test set.
- Adds additional parameters to configuration file
- Cleans up configuration and slurm files

KimballJardine committed:
Revert "Delete train_model_lightning.py"

This reverts commit d37d1be2f8539f1c2d3f568f5c1c97a540fd965e.

JayOrten committed:
Revert "remove dataset.py, lightning class"

This reverts commit 63740617396d57c909540b5475e6c1f420f226d7.

KimballJardine committed:
Revert "remove dataset.py, lightning class"

This reverts commit 63740617396d57c909540b5475e6c1f420f226d7.

JayOrten committed:
Merge branch 'main' into implement_yaml

JayOrten committed:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

JayOrten committed:
refactor code to use yaml file

JayOrten committed:
organize imports in train

JayOrten committed:
refactor dataset class download paths

JayOrten committed:
cleanup dead code

JayOrten committed:
add checkpoint path

JayOrten committed:
update model checkpoint filename. I replaced num_val_runs, with just the loss, not sure how to keep track of validation count, but probably not relevant

JayOrten committed:
add new parameters and sc check

JayOrten committed:
enable lr scheduler and add gamma

JayOrten committed:
remove dead model initialization. Also, remove fsdp, because lightning handles that

JayOrten committed:
add user slurm folder

JayOrten committed:
Merge remote-tracking branch 'origin/main' into implement_lightning.kachow!

JayOrten committed:
add config changes

JayOrten committed:
misc changes

JayOrten committed:
Fix multiline string

JayOrten committed:
fix model type parsing

JayOrten committed:
fix parameter name, fix torch compile

JayOrten committed:
fix learning rate parameter, add time stamp

JayOrten committed:
fix datetime import

JayOrten committed:
update default slurm script

JayOrten committed:
remove compile because it doesn't work with our setup

JayOrten committed:
add todo

JayOrten committed:
add todos

Commit:
Added a ton of stuff, i created generate_specific_text, inference_test.py, and models.py.  I also made a slurm script that runs generate specific text

DrewGalbraith committed:
correct dataset path in yaml template
 Please enter the commit message for your changes. Lines starting

JayOrten committed:
Consolidate model classes and get generate working

JayOrten committed:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

JayOrten committed:
remove model code from train, rename files

JayOrten committed:
clean up imports

JayOrten committed:
update todo and remove huggingface model.py

JayOrten committed:
implement lr scheduler

JayOrten committed:
Get test set inference working, update model type checking

JayOrten committed:
remove unused parameters from config

JayOrten committed:
remove extra heads parameter

JayOrten committed:
add slurm scripts

JayOrten committed:
update readme

JayOrten committed:
remove whitespace

JayOrten committed:
remove unused head parameter

nprisbrey commented on TODO.md:
I assume we're working towards deleting this file, correct?

JayOrten replied:
Yes

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -23,7 +23,7 @@ dataset_feature: "text"
 dataset_name: "wikitext"
 # HuggingFace Dataset Subset (str)
-dataset_subset: "wikitext-2-v1"
+dataset_subset: "wikitext-103-raw-v1"
```
Is there a reason that we're changing from the lightest dataset to get someone up and running or is this just residual code?

JayOrten replied:
Nope, just residual. Changed

nprisbrey commented on slurm/slurm_template.sh:
```diff
@@ -1,41 +1,17 @@
 #SBATCH --output=%x_%j.out
 
 # Set the max number of threads to use for programs using OpenMP. Should be <= ppn. Does nothing if the program doesn't use OpenMP.
 export OMP_NUM_THREADS=$SLURM_CPUS_ON_NODE
```
With our new changes of moving to different QOS and changing our requirements, does it still make sense to be exporting this OpenMP variable?

JayOrten replied:
Not necessarily, we likely weren't using OpenMP anyways for anything, QOS wouldn't change that. This is here simply because ORC recommends it, but we could remove it.

JayOrten replied:
Update on this: removing this line can result in strange behaviour.

nprisbrey committed:
Make run_eval.sh Path Match Other Slurm Scripts

nprisbrey committed:
Rename Slurm Training Script For Clarity

"train.sh" was too generic, therefore it is now "train_model.sh".

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -44,12 +44,41 @@ models_path: "/YOUR/PATH/HERE/data/models"
   - 0.2
   - 0.1
 
+# Device Configuration
+device: "cuda" # Options: "cpu", "cuda"
+# num_nodes (int): Number of nodes
+num_nodes: 1
+# Number of GPUs (int)
+num_devices: 1 
+# strategy (string): Distributed strategy for training. You probably don't need to change this
+strategy: "ddp" # Options: "ddp", "ddp_spawn". 
+
+# Training Configuration
+rand_seed: 42
+# epochs (int): Number of epochs to train for.
+epochs: 1
+# lr (float): Learning rate of model to train.
+learning_rate: 0.001 # lr
+# save_top_k (int): Number of best models to save.
+save_top_k: 3
+# val_check_interval (float): How often to validate as a fraction of an epoch.
+val_check_interval: 0.5
+# use_slurm (bool): Whether to use SLURM for training.
+use_slurm: true
+# gamma (float): Learning rate scheduler gamma.
+gamma: 0.85
+# accumulate_grad_batches (int): Accumulate gradients over n batches.
+accumulate_grad_batches: 1
+
```
Is there a reason that we're pulling this out of the "MODEL TRAINING CONFIGURATION" section below?

JayOrten replied:
I would argue yes. Model Training Configuration should have the parameters for actual model configuration, i.e., layers, width, etc. These parameters are settings for the devices and the hyperparameters, such as LR, epochs, etc. I would say separating them makes it easier to parse through. These also need to be reorganized a bit.

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -44,12 +44,41 @@ models_path: "/YOUR/PATH/HERE/data/models"
 tboard_path: ~
 
+# Checkpoint Path: path to checkpoint to load for generation/inference
+checkpoint_path: ~
```
If we are to implement a `checkpoint_path` argument, shouldn't this be implemented everywhere (ex: the normal `train_model.py`)?

JayOrten replied:
The purpose of this parameter is that, once you've trained a model and want to do inference or generation, you put into this field the path to the checkpoint that you want to use for those tasks. This isn't used in the training scripts.

nprisbrey commented on train_model_lightning.py:
```diff
@@ -0,0 +1,188 @@
+        
+    config = Struct(**config)
+
+    start_time = time.time()
+
+    model, test_loss = train_model(config)
+
+    end_time = time.time()
```
What is the point of these times? It doesn't seem that they're used anywhere....

JayOrten replied:
removed

nprisbrey commented on train_model_lightning.py:
```diff
@@ -0,0 +1,188 @@
+    print(f"Saving model files in {model_dir}")
+
+    # Initialize checkpoints directory
+    checkpoints_dir = model_dir / "checkpoints"
```
Seems like this should be refactored to use the new config parameter.

JayOrten replied:
See other comment on config parameter

nprisbrey reviewed (commented):
Some comments on the current status of things. Let it be recognized that I may have just been commenting on work that was already planning on being done but hasn't come to pass yet.

JayOrten committed:
change template default dataset

JayOrten committed:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

JayOrten committed:
remove time in train

JayOrten committed:
fix returned loss value

JayOrten committed:
organize template config

JayOrten committed:
inference works

JayOrten committed:
update todo

Commit:
Added script to run inference

Commit:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

JayOrten committed:
update yaml with example tboard path

JayOrten committed:
reorganize imports

JayOrten committed:
add tensorboard saving path logic and logger to trainer

Commit:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

JayOrten committed:
remove extra epochs parameter

Commit:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

Added Inference.sh script to run inference

JayOrten committed:
update train model script to allow requeing

JayOrten committed:
make it okay for directories to exist, important for multiple processes?

JayOrten committed:
remove todo because it's basically done

JayOrten commented:
Based on the state of the code and results from test runs I feel comfortable opening this up

KimballNJardine commented on train_model_lightning.py:
```diff
@@ -0,0 +1,187 @@
+
+# PyTorch Lightning
+from pytorch_lightning import Trainer
+from pytorch_lightning.plugins.environments import SLURMEnvironment
```
Alphabetize

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -59,35 +92,20 @@ activation_dropout: 0.0
-num_devices: 1
-# Random Seed (int): For reproducibility
-rand_seed: 42
+model_type: retnet # Choices: "retnet", "transformer"
```
Is there a reason why we're making strings no longer in double quotes (ex: `"retnet"`)?

JayOrten replied:
Technically quotes are unnecessary in yaml, but changed for consistency

nprisbrey committed:
Standardize Config Comment Formatting

nprisbrey committed:
Arrange Config Variables Alphabetically

nprisbrey commented on dataset.py:
It seems that most typing hints and all documentation strings are missing from this file.

nprisbrey commented on dataset.py:
```diff
@@ -0,0 +1,40 @@
+from datasets import load_dataset
+
+class DataModule(LightningDataModule):
+    def __init__(self, config, num_workers=0):
```
What's stopping us from putting `num_workers` into the config file like everything else? This could also maintain consistency with `train_model_lightning.py` and remove a constant from both files.

JayOrten replied:
Added. There wasn't any reason why, I just haven't experimented with that parameter as much

nprisbrey committed:
Refactor PyTorch Lightning Dataset Preparation

Adjust code to better match recommendations and styling given at
"https://lightning.ai/docs/pytorch/stable/data/datamodule.html".

nprisbrey commented on slurm/train_model.sh:
```diff
@@ -0,0 +1,19 @@
+
+# LOAD MODULES, INSERT CODE, AND RUN YOUR PROGRAMS HERE
+mamba activate retnet
+srun python3 ../../train_model_lightning.py ../../configs/user_configs/YOUR_CONFIG_HERE.yaml
```
Just double checking: is calling `srun` here for a job step with Slurm done on purpose for the requeueing logic or is this extra and could be removed?

JayOrten replied:
Yes, without `srun` the script will hang while initializing devices

nprisbrey commented on models.py:
Another reminder of updating Doc Strings.

JayOrten replied:
added, let me know if I missed anything that should be explained more fully

nprisbrey commented on train_model_lightning.py:
```diff
@@ -0,0 +1,187 @@
+# better performance while training if hardware is capable
+torch.backends.cuda.matmul.allow_tf32 = True
+
+class CustomCheckpoint(ModelCheckpoint):
```
Can we rename this possibly to `CustomModelCheckpoint` or something similar to be more descriptive?

JayOrten replied:
changed

nprisbrey commented on configs/template_config.yaml:
```diff
@@ -59,35 +92,20 @@ activation_dropout: 0.0
 ffn_dim: 12
-# Fully Sharded Data Parallel (bool): Whether to shard Module parameters across
-# data parallel workers or not (with the FairScale library)
-fsdp: false
```
The normal `train_model.py` script no longer runs because `fsdp` and `validation_frequency` are no longer included in the config file. How are we expecting to resolve this?

JayOrten replied:
I will add them back in. However, we should consider whether or not we want to maintain both `train_model.py` and `train_model_lightning.py`, because the latter should be able to do everything the former is doing, while enabling multi-device training.

nprisbrey replied:
I think that, as you mention, we should just drop the original `train_model.py` script in favor of Lightning. Thoughts @JayOrten and @KimballNJardine?

JayOrten replied:
I'm okay with that

KimballNJardine replied:
That is fine with me.

nprisbrey commented on slurm/train_model.sh:
Do we have a reasoning for why we only have a `train_model_lightning.py` script in the `slurm` directory and only have a `train_model.py` script in the `scripts` directory? `inference.sh` and `generate.sh` don't seem to have any equivalent in the `scripts` directory either.

JayOrten replied:
Nope, I've just only been testing through slurm. Added

nprisbrey committed:
Standardize Model Construction

nprisbrey committed:
Fix Broken Dependency For Non-Lightning Training

nprisbrey committed:
Get Lightning Training Working

nprisbrey reviewed (changes requested):
These are some preliminary thoughts @JayOrten. Let me know what you think and then I'll finish looking over the code. I already made some changes myself but would also like to see what your thoughts are on changing more.

JayOrten committed:
add quotations

JayOrten committed:
add fsdp parameter for hf

JayOrten committed:
change val check parameter name in train model

JayOrten committed:
add num workers parameter

JayOrten committed:
rename custom checkpoint

JayOrten committed:
add/update slurm and bash scripts

JayOrten committed:
update docstring in dataset

JayOrten committed:
add doc strings, type hints

nprisbrey committed:
Format And Make Executable Scripts

nprisbrey committed:
Fix Bug Related To Transition To val_check_interval

We need to change the equation for triggering validation because
validation_frequency was a whole number, whereas val_check_interval is a
float.

nprisbrey committed:
Start Removal Of PyTorch-only Code

JayOrten committed:
add early stopping

nprisbrey committed:
Finish Removal Of PyTorch-only Code

nprisbrey committed:
Update dataset.py Documentation Strings

nprisbrey committed:
Drop The fsdp Config Parameter

nprisbrey committed:
Drop Support For Config Objects From Strings

nprisbrey committed:
Touch Up generation.py And Add gen_len

Add new parameter to the configuration file to prevent hard-coding the
generation length.

nprisbrey committed:
Return Best Model Instead Of Latest Model During Training

nprisbrey committed:
Update grid_search.py Documentation String

nprisbrey committed:
Organize and Comment inference.py

nprisbrey committed:
rganize and Comment models.py

KimballJardine committed:
reorganize package imports

KimballJardine committed:
Merge remote-tracking branch 'refs/remotes/origin/implement_lightning.kachow!' into implement_lightning.kachow!

KimballJardine committed:
Fix order of imports in inference.py

KimballJardine committed:
Clean up new lines, ' --> " in run_eval.py

nprisbrey committed:
Standardize Script Formats And Permissions

Note that `scripts/run_eval.sh` currently won't run until Kimball
finishes touching up `run_eval.py`.

nprisbrey committed:
Add Back EOF For .gitignore

nprisbrey committed:
Simplify train_model.py When Run Directly

KimballJardine committed:
Move run_eval.py to YAML format.

KimballJardine committed:
Move input strings to config file.

KimballJardine committed:
Remove uneeded space.

KimballJardine committed:
More cleanup of yaml file, fix run_eval.sh files

KimballJardine committed:
Add types to variables in yaml.

nprisbrey committed:
Organizing template_config.yaml

KimballJardine committed:
Alphabetize YAML config file.

KimballJardine committed:
Clean up error checking handling in run_eval.py

KimballJardine committed:
Merge branch 'implement_lightning.kachow!' of github.com:DRAGNLabs/301r_retnet into implement_lightning.kachow!

nprisbrey committed:
Touchups For run_eval.py

KimballJardine committed:
Give better instructions for paths for run_eval.py

nprisbrey committed:
Fix Errors In Configuration Template

nprisbrey committed:
Fix Bugs Preventing Scripts From Running

nprisbrey reviewed (approved):
Looks good! All scripts in the `scripts` directory function as expected. I haven't tested any of the Slurm scripts myself, though they run on the same underlying code and I believe they should be easy to use for this weekend.

KimballNJardine reviewed (approved):
Looks good. Approved.
